lib/segment/src/index/hnsw_index/graph_links.rs

```diff
@@ -50,44 +55,121 @@ links offset = level_offsets[level] + offsets[reindex[point_id]]
 
 const HEADER_SIZE: usize = 64;
 
-#[derive(Clone, Debug, Default)]
+#[derive(Clone, Debug)]
 struct GraphLinksFileInfo {
     point_count: usize,
     reindex_start: usize,
     links_start: usize,
     offsets_start: usize,
     offsets_end: usize,
+    compression: Option<CompressionInfo>,
+}
+
+#[derive(Clone, Debug)]
+struct CompressionInfo {
+    m: usize,
+    m0: usize,
+    bits_per_unsorted: u8,
 }
 
 #[derive(AsBytes, FromBytes, FromZeroes)]
 #[repr(C)]
-struct GraphLinksFileHeader {
+struct HeaderV0 {
     point_count: u64,
     levels_count: u64,
-    total_links_len: u64,
-    total_offsets_len: u64,
-    offsets_padding: u64, // 0 or 4
+    total_links_count: u64,
+    total_offset_count: u64,
+    /// Either 0 or 4.
+    offsets_padding_bytes: u64,
+    zero_padding: [u8; 24],
 }
 
+/// File header for the compressed format.
+#[derive(AsBytes, FromBytes, FromZeroes)]
+#[repr(C)]
+struct HeaderV1 {
+    point_count: U64,
+    /// Should be [`HEADER_VERSION_1`].
+    ///
+    /// Deliberately placed at the same offset as [`HeaderV0::levels_count`] and
+    /// set to an impossibly large number to make old Qdrant versions fail fast
+    /// when trying to read the new format.
+    version: U64,
```

Comment:
I want an old Qdrant (≤1.12) to explicitly crash if it try to read a compressed file. If we place a new field to the end, it would just ignore it as it doesn't know about it.

While links compression is a hidden experimental feature that is not enabled by default, we won't bump the segment version soon.

Reply:
As we discussed separately, old qdrant versions will throw the error in this approach. But the error will be `segfault`, it does not say that user tries to open new links with older qdrant. I don't remember why we cannot increase the segment version in this case, can you remind please complains about the segment version update?